File: src/main/parsers/languages.ts
// Language definitions: extension -> language id, with comment syntax.
// Inspired by cloc; minimal but covers common languages.

export interface LangDef {
  id: string;
  line: string[]; // line comment markers
  block: Array<[string, string]>; // [start, end]
  string?: Array<[string, string]>; // simple string delimiters to skip
}

const QUOTED_STRINGS: Array<[string, string]> = [['"', '"'], ["'", "'"]];

const C_LIKE: Pick<LangDef, 'line' | 'block' | 'string'> = {
  line: ['//'],
  block: [['/*', '*/']],
  string: [...QUOTED_STRINGS, ['`', '`']],
};

const WEB_COMPONENT: Pick<LangDef, 'line' | 'block' | 'string'> = {
  line: ['//'],
  block: [['<!--', '-->'], ['/*', '*/']],
  string: [...QUOTED_STRINGS, ['`', '`']],
};

export const LANGUAGES: Record<string, LangDef> = {
  ts: { id: 'TypeScript', ...C_LIKE },
  tsx: { id: 'TSX', ...C_LIKE },
  js: { id: 'JavaScript', ...C_LIKE },
  jsx: { id: 'JSX', ...C_LIKE },
  mjs: { id: 'JavaScript', ...C_LIKE },
  cjs: { id: 'JavaScript', ...C_LIKE },
  json: { id: 'JSON', line: [], block: [] },
  c: { id: 'C', ...C_LIKE },
  h: { id: 'C/C++ Header', ...C_LIKE },
  cpp: { id: 'C++', ...C_LIKE },
  cc: { id: 'C++', ...C_LIKE },
  hpp: { id: 'C++ Header', ...C_LIKE },
  java: { id: 'Java', ...C_LIKE },
  kt: { id: 'Kotlin', ...C_LIKE },
  swift: { id: 'Swift', ...C_LIKE },
  go: { id: 'Go', ...C_LIKE },
  rs: { id: 'Rust', ...C_LIKE },
  cs: { id: 'C#', ...C_LIKE },
  scala: { id: 'Scala', ...C_LIKE },
  php: { id: 'PHP', line: ['//', '#'], block: [['/*', '*/']], string: C_LIKE.string },
  py: { id: 'Python', line: ['#'], block: [['"""', '"""'], ["'''", "'''"]], string: QUOTED_STRINGS },
  rb: { id: 'Ruby', line: ['#'], block: [['=begin', '=end']], string: QUOTED_STRINGS },
  sh: { id: 'Shell', line: ['#'], block: [], string: QUOTED_STRINGS },
  bash: { id: 'Shell', line: ['#'], block: [], string: QUOTED_STRINGS },
  zsh: { id: 'Shell', line: ['#'], block: [], string: QUOTED_STRINGS },
  yml: { id: 'YAML', line: ['#'], block: [], string: QUOTED_STRINGS },
  yaml: { id: 'YAML', line: ['#'], block: [], string: QUOTED_STRINGS },
  toml: { id: 'TOML', line: ['#'], block: [], string: QUOTED_STRINGS },
  ini: { id: 'INI', line: [';', '#'], block: [], string: QUOTED_STRINGS },
  sql: { id: 'SQL', line: ['--'], block: [['/*', '*/']], string: QUOTED_STRINGS },
  html: { id: 'HTML', line: [], block: [['<!--', '-->']] },
  htm: { id: 'HTML', line: [], block: [['<!--', '-->']] },
  xml: { id: 'XML', line: [], block: [['<!--', '-->']] },
  css: { id: 'CSS', line: [], block: [['/*', '*/']] },
  scss: { id: 'SCSS', ...C_LIKE },
  less: { id: 'Less', ...C_LIKE },
  md: { id: 'Markdown', line: [], block: [] },
  vue: { id: 'Vue', ...WEB_COMPONENT },
  svelte: { id: 'Svelte', ...WEB_COMPONENT },
  lua: { id: 'Lua', line: ['--'], block: [['--[[', ']]']] },
  pl: { id: 'Perl', line: ['#'], block: [['=pod', '=cut']] },
  r: { id: 'R', line: ['#'], block: [] },
  dart: { id: 'Dart', ...C_LIKE },
  ex: { id: 'Elixir', line: ['#'], block: [] },
  exs: { id: 'Elixir', line: ['#'], block: [] },
  elm: { id: 'Elm', line: ['--'], block: [['{-', '-}']] },
  hs: { id: 'Haskell', line: ['--'], block: [['{-', '-}']] },
  clj: { id: 'Clojure', line: [';'], block: [] },
  vim: { id: 'Vim Script', line: ['"'], block: [] },
  dockerfile: { id: 'Dockerfile', line: ['#'], block: [] },
  makefile: { id: 'Makefile', line: ['#'], block: [] },
};

const SPECIAL_FILENAMES: Record<string, string> = {
  Dockerfile: 'dockerfile',
  Makefile: 'makefile',
  GNUmakefile: 'makefile',
};

export function detectLang(relPathOrName: string): { ext: string; lang: LangDef | null; langId: string } {
  const base = relPathOrName.split(/[\\/]/).pop() || relPathOrName;
  if (SPECIAL_FILENAMES[base]) {
    const k = SPECIAL_FILENAMES[base];
    return { ext: k, lang: LANGUAGES[k], langId: LANGUAGES[k].id };
  }
  const dot = base.lastIndexOf('.');
  if (dot < 0) return { ext: '', lang: null, langId: 'Other' };
  const ext = base.slice(dot + 1).toLowerCase();
  const lang = LANGUAGES[ext] || null;
  return { ext, lang, langId: lang ? lang.id : 'Other' };
}

File: src-tauri/src/parsers/languages.rs
#[derive(Clone, Debug)]
pub struct LangDef {
    pub id: &'static str,
    pub line: &'static [&'static str],
    pub block: &'static [(&'static str, &'static str)],
    pub string: &'static [(&'static str, &'static str)],
}

const QS: &[(&str, &str)] = &[("\"", "\""), ("'", "'")];
const C_LINE: &[&str] = &["//"];
const C_BLOCK: &[(&str, &str)] = &[("/*", "*/")];
const C_STR: &[(&str, &str)] = &[("\"", "\""), ("'", "'"), ("`", "`")];
const WEB_BLOCK: &[(&str, &str)] = &[("<!--", "-->"), ("/*", "*/")];
const PY_BLOCK: &[(&str, &str)] = &[("\"\"\"", "\"\"\""), ("'''", "'''")];

fn c_like(id: &'static str) -> LangDef {
    LangDef { id, line: C_LINE, block: C_BLOCK, string: C_STR }
}

pub fn detect_lang(rel_path: &str) -> (String, Option<LangDef>, String) {
    let base = rel_path.rsplit(['/', '\\']).next().unwrap_or(rel_path);
    let special = match base {
        "Dockerfile" => Some(("dockerfile", LangDef { id: "Dockerfile", line: &["#"], block: &[], string: QS })),
        "Makefile" | "GNUmakefile" => Some(("makefile", LangDef { id: "Makefile", line: &["#"], block: &[], string: QS })),
        _ => None,
    };
    if let Some((ext, lang)) = special {
        return (ext.to_string(), Some(lang.clone()), lang.id.to_string());
    }
    let ext = match base.rsplit_once('.') {
        Some((_, e)) => e.to_lowercase(),
        None => return (String::new(), None, "Other".into()),
    };
    let lang = match ext.as_str() {
        "ts" => Some(c_like("TypeScript")),
        "tsx" => Some(c_like("TSX")),
        "js" | "mjs" | "cjs" => Some(c_like("JavaScript")),
        "jsx" => Some(c_like("JSX")),
        "json" => Some(LangDef { id: "JSON", line: &[], block: &[], string: &[] }),
        "c" => Some(c_like("C")),
        "h" => Some(c_like("C/C++ Header")),
        "cpp" | "cc" => Some(c_like("C++")),
        "hpp" => Some(c_like("C++ Header")),
        "java" => Some(c_like("Java")),
        "kt" => Some(c_like("Kotlin")),
        "swift" => Some(c_like("Swift")),
        "go" => Some(c_like("Go")),
        "rs" => Some(c_like("Rust")),
        "cs" => Some(c_like("C#")),
        "scala" => Some(c_like("Scala")),
        "php" => Some(LangDef { id: "PHP", line: &["//", "#"], block: C_BLOCK, string: C_STR }),
        "py" => Some(LangDef { id: "Python", line: &["#"], block: PY_BLOCK, string: QS }),
        "rb" => Some(LangDef { id: "Ruby", line: &["#"], block: &[("=begin", "=end")], string: QS }),
        "sh" | "bash" | "zsh" => Some(LangDef { id: "Shell", line: &["#"], block: &[], string: QS }),
        "yml" | "yaml" => Some(LangDef { id: "YAML", line: &["#"], block: &[], string: QS }),
        "toml" => Some(LangDef { id: "TOML", line: &["#"], block: &[], string: QS }),
        "ini" => Some(LangDef { id: "INI", line: &[";", "#"], block: &[], string: QS }),
        "sql" => Some(LangDef { id: "SQL", line: &["--"], block: C_BLOCK, string: QS }),
        "html" | "htm" => Some(LangDef { id: "HTML", line: &[], block: &[("<!--", "-->")], string: &[] }),
        "xml" => Some(LangDef { id: "XML", line: &[], block: &[("<!--", "-->")], string: &[] }),
        "css" => Some(LangDef { id: "CSS", line: &[], block: C_BLOCK, string: &[] }),
        "scss" => Some(c_like("SCSS")),
        "less" => Some(c_like("Less")),
        "md" => Some(LangDef { id: "Markdown", line: &[], block: &[], string: &[] }),
        "vue" | "svelte" => Some(LangDef {
            id: if ext == "vue" { "Vue" } else { "Svelte" },
            line: C_LINE, block: WEB_BLOCK, string: C_STR,
        }),
        "lua" => Some(LangDef { id: "Lua", line: &["--"], block: &[("--[[", "]]")], string: &[] }),
        "dart" => Some(c_like("Dart")),
        "ex" | "exs" => Some(LangDef { id: "Elixir", line: &["#"], block: &[], string: &[] }),
        "hs" | "elm" => Some(LangDef { id: if ext == "hs" { "Haskell" } else { "Elm" }, line: &["--"], block: &[("{-", "-}")], string: &[] }),
        _ => None,
    };
    let lang_id = lang.as_ref().map(|l| l.id.to_string()).unwrap_or_else(|| "Other".into());
    (ext, lang, lang_id)
}
